### ISSUE TITLE
Secure connections strings configuration API is not clear

### ISSUE THREAD

**Issue by SeanFeldman** [Particular/NServiceBus.AzureStorageQueues#119]
Connects to https://github.com/Particular/PlatformDevelopment/issues/871

Going through configuration API and reading ["Securing connections strings to storage accounts"](http://docs.particular.net/nservicebus/azure-storage-queues/configuration#connection-strings-securing-connection-strings-to-storage-accounts) I've realized that I don't understand how the API should be used. The 2 almost identical code snippets are very confusing:

```c#
var transport = endpointConfiguration.UseTransport<AzureStorageQueueTransport>();
transport
    .ConnectionString("default_connection_string")
    .UseAccountNamesInsteadOfConnectionStrings(mapping =>
    {
        mapping.MapLocalAccount("default");
        mapping.MapAccount("another","another_connection_string");
    });
```
and
```c#
var transport = endpointConfiguration.UseTransport<AzureStorageQueueTransport>();
transport
    .ConnectionString("another_connection_string")
    .UseAccountNamesInsteadOfConnectionStrings(mapping =>
    {
        mapping.MapLocalAccount("another");
        mapping.MapAccount("default", "default_connection_string");
    });
```
The questions I had were:

1. Why the two identical code snippets?
1. What's the affect on configuration if `.ConnectionString()` is used? Is that a `default` connections string?
1. What is `MapLocalAccount()`?
1. How `.ConnectionString()` and `.UseAccountNamesInsteadOfConnectionStrings()` with mapping affect default connection string?
1. How `mapping.MapAccount("default", "default_connection_string");` relates to `ConnectionString()`?

**Comment by Scooletz** [comment]
Let me try to answer your questions. But first, let me remark that maybe this documentation piece should be moved to http://docs.particular.net/nservicebus/azure-storage-queues/multi-storageaccount-support
 
> 1. Why the two identical code snippets?

Every snippet represents a separate endpoint using a different account for its input queue.

> 2. What's the affect on configuration if .ConnectionString() is used? Is that a default connections string?

Yes, that's the default connection string.

> 3. What is MapLocalAccount(name)

This maps the default, local connection string with the name, so when sending to another account, the name will be used for the `ReplyTo` header.

> 4. How .ConnectionString() and .UseAccountNamesInsteadOfConnectionStrings() with mapping affect default connection string?

Mapping does not change the connection string value. If you set the connection string with `.ConnectionString()` it is set. Mapping is only for cross account sends.

> 5. How mapping.MapAccount("default", "default_connection_string"); relates to ConnectionString()

Let's consider an example, where in the account specified by `default_connection_string` there is a `receiver` queue. The mapping tells only, that if a message is sent to `receiver@default`, the `default_connection_string` should be used to access the account and access a `receiver` queue.

**Comment by Scooletz** [comment]
I hope this answers your questions. Nevertheless, if a maintainer of ASQ sees a problem in docs, there's definitely something lacking. I'm going to write these explanations in a PR to docs. If you have anything more, especially if it's connected with the feature itself, not doco only, please ping me.

**Comment by SeanFeldman** [comment]
Conversation is taking place on PR https://github.com/Particular/docs.particular.net/issues/1526

**Comment by Scooletz** [comment]
Related to ASB, but describing its API behavior: https://github.com/Particular/docs.particular.net/pull/1640

**Comment by yvesgoeleven** [comment]
The API as we ended up with in ASB:

NamespacePartitioning().AddNamespace(name, cstring)
NamespaceRouting().AddNamespace(name, cstring)
ConnectionString(cstring)
DefaultNamespaceName(name)
UseNamespaceNamesInsteadOfConnectionstring()

**Comment by Scooletz** [comment]
For now, as ASQ does not implement `partitioning`, I'm aiming at changing the API to:
- `NamespaceRouting().AddNamespace(name, cstring)`
- `ConnectionString(cstring)`
- `DefaultNamespaceName(name)`
- `UseNamespaceNamesInsteadOfConnectionstring()`

**Comment by dvdstelt** [comment]
I really think we should have API changes like this, especially aligning APIs, on a broader scale. I would even go so far to say we need an API maintainer group.
I've already heard so many opinions on the API and whether or not we should align them, that I don't think maintainers of a transport or persister can make the adjustment on their own. Especially since public API is so important and we change things too often.

First of all, where does 'Namespace' suddenly come from? I do get ASB namespaces, but I don't get ASQ namespaces?
Why is `DefaultNamespaceName` not within `NamespaceRouting`, as `AddNamespace` is.
And I think `UseNamespaceNamesInsteadOfConnectionstring` is way, way too long. I'd rather have `EnableNamespaces`

**Comment by SeanFeldman** [comment]
@dvdstelt I'm sure @Scooletz meant to use _Account_ and not _Namespace_ and it was just a copy/paste issue :smiley: 

Now as for the personal preferences, while you disagree with `UseNamespaceNamesInsteadOfConnectionStrings`, I find it much more descriptive and revealing than `EnableNamespaces`. The latter is somewhat cryptic. Enable namespaces, does it mean by default they are disabled? Aren't namespaces always used (or at least one)?

**Comment by Scooletz** [comment]
Yes @dvdstelt & @SeanFeldman , it was a copy paste - my bad. The API of the secure connection strings and using multiple accounts has already been changed multiple times. We need to close this as well as close the docs PR  Particular/docs.particular.net#1526 release and imo it should be done right now.
I think that aligning to an already existing solution in NSB is slightly better than providing a custom API for ASQ itself.
I think @dvdstelt that we've got this group already. They are core maintainers which actually provide seams for all the implementations, still, leaving some space for transports/persistence.

**Comment by melkio** [comment]
@dvdstelt `DefaultNamespaceName` is related to `UseNamespaceNamesInsteadOfConnectionString`: it's to cover the scenario for which a user enables `securing connection string` and they configure `connection string` through `ConnectionString` extension method without the possibility to specify namespace/account name. 
Given that I think `DefaultNamespaceName` api should be at the same level of `UseNamespaceNamesInsteadOfConnectionString`

**Comment by yvesgoeleven** [comment]
@dvdstelt `DefaultNamespaceName` is also relevant for partitioning, not just routing. Partitioning hasn't been introduced in ASQ yet, but that's just a matter of time. better to put it a level up in the API

**Comment by Scooletz** [comment]
Solved in #134
